pkg/lib/import-json.ts

```diff
@@ -0,0 +1,262 @@
+         ]
+       }
+
+    The "name" field for Håland is not a string, and it is
+    mandatory. An error will be logged and the whole Håland entry is
+    omitted from "players".
+
+      {
+         "name": "ManCity",
+         "players": "TBD"
+       }
+
+    The "players" field is not an array, and it is mandatory. The
+    import_Team function will raise a ValidationError exception.
+ */
+
+/* WRITING IMPORTER FUNCTIONS
+
+   The process of writing a importer function for a given TypeScript
+   interface is pretty mechanic, and could well be automated.
+
+   For example, these are the functions for the Player and Team
+   interfaces from above.
+
+       interface Player {
+          name: string;
+          age?: number;
+       }
+
+       function import_Player(val: JsonValue): Player {
+           const obj = import_json_object(val);
+           const res: Player = {
+               name: import_mandatory(obj, "name", import_string),
+           };
+           import_optional(res, obj, "age", import_number);
+           return res;
+       }
+
+       interface Team {
+          name: string;
+          players: Player[];
+       }
+
+       function import_Team(val: JsonValue): Team {
+           const obj = import_json_object(val);
+           const res: Team = {
+               name: import_mandatory(obj, "name", import_string),
+               players: import_mandatory(obj, "players", v => import_array(v, import_Player))
+           };
+           return res;
+       }
+
+   More examples can be found in "pkg/shell/manifests.ts".
+ */
+
+class ValidationError extends Error { }
+
+const validation_path: string[] = [];
+
+function with_validation_path<T>(p: string, func: () => T): T {
+    validation_path.push(p);
+    try {
+        return func();
+    } finally {
+        validation_path.pop();
+    }
+}
+
+function validation_error(msg: string): never {
+    console.error(`JSON validation error for ${validation_path.join("")}: ${msg}`);
+    throw new ValidationError();
+}
+
+export function import_string(val: JsonValue): string {
+    if (typeof val == "string")
+        return val;
+    validation_error(`Not a string: ${JSON.stringify(val)}`);
+}
+
+export function import_number(val: JsonValue): number {
+    if (typeof val == "number")
+        return val;
+    validation_error(`Not a number: ${JSON.stringify(val)}`);
+}
+
+export function import_boolean(val: JsonValue): boolean {
+    if (typeof val == "boolean")
+        return val;
+    validation_error(`Not a boolean: ${JSON.stringify(val)}`);
+}
+
+export function import_json_object(val: JsonValue): JsonObject {
+    if (!!val && typeof val == "object" && val.length === undefined)
+        return val as JsonObject;
+    validation_error(`Not an object: ${JSON.stringify(val)}`);
+}
+
+export function import_json_array(val: JsonValue): JsonValue[] {
+    if (!!val && typeof val == "object" && val.length !== undefined)
+        return val as JsonValue[];
```

Comment:
You should try to get this working without resorting to unsafe casts.  The compiler has a fairly impressive set of narrowing mechanisms...

See also https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray